infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,380 @@
+            Xai(xai).mint(esXaiBurnFoundationRecipient, foundationXaiAmount);
+        }
+
+        // emit event of the redemption
+        emit RedemptionCompleted(msg.sender, index);
+    }
+
+    /**
+     * @dev Function to get the redemption request at a given index.
+     * @param account The address to query.
+     * @param index The index of the redemption request.
+     * @return The redemption request.
+     */
+    function getRedemptionRequest(address account, uint256 index) public view returns (RedemptionRequestExt memory) {
+        return _extRedemptionRequests[account][index];
+    }
+
+    /**
+     * @dev Function to get the count of redemption requests for a given address.
+     * @param account The address to query.
+     * @return The count of redemption requests.
+     */
+    function getRedemptionRequestCount(address account) public view returns (uint256) {
+        return _extRedemptionRequests[account].length;
+    }
+
+    /**
+     * @dev Function to get the count of redemption requests for a given address.
+     * @param number The amount to update the basepoints.
+     */
+    function updateFoundationBasepoints(uint256 number) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(number <= 1000, "Invalid basepoints");
+        esXaiBurnFoundationBasePoints = number;
+        emit FoundationBasepointsUpdated(number); 
+    }
+
+    /**
+     * @dev Function to set the max keys allowed for a non kyc
+     * @param newMax The new max keys allowed.
+     */
+    function setMaxKeysNonKyc(uint256 newMax) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        maxKeysNonKyc = newMax;
+    }
+
+    /**
+    * Converts redemptions in process to vouchers for a given list of accounts and indices.
+    * 
+    * This function processes redemptions that are currently in process and converts them into vouchers. 
+    * It requires that redemptions be paused, and validates that the length of `accounts` matches the length of `indices`.
+    * For each redemption request, it checks if the request is not completed and if the voucher has not been issued.
+    * If both conditions are met, it marks the voucher as issued, updates the total pending redemptions, 
+    * and transfers the corresponding amount to the account.
+    * 
+    * @param accounts - The list of account addresses for which redemptions are to be converted.
+    * @param indices - The list of indices corresponding to redemption requests for each account.
+    * Will throw an error if redemptions are active or if the lengths of `accounts` and `indices` do not match.
+    * @dev Only callable by an account with the `DEFAULT_ADMIN_ROLE`.
+    */
+    function convertRedemptionsInProcess(address[] calldata accounts, uint256[][] calldata indices) nonReentrant external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(!_redemptionActive, "Redemptions must be paused to convert");
```

Comment:
Why do we need to pause ?

Reply:
Because starting a new redemption will require us to know the total pending redemptions in process in order to make sure they have enough esXai to start the claim.
We won't know the total pending redemptions in process until all of the conversions have been completed.

Reply:
You would not be able to start a new redemption if you don't have the esXai in your wallet.
So we should be safe, after the upgrade you get to keep the esXai, any request that has not been migrated yet will still have your esXai on the contract, so the total pending should be correct. Let me know if Im missing something here.